### PR TITLE
+ extunix 0.3.1

### PR DESCRIPTION
**Pull request by ygrek**
follow-up to https://github.com/ocaml/opam-repository/pull/18478 with compatibility maintained

### CODE REVIEW

File: packages/extunix/extunix.0.3.1/opam [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+bug-reports: "https://github.com/ygrek/extunix/issues"
+depends: [
+  "dune" {>= "2.2"}
+  "dune-configurator" {build}
+  "ppxlib" {>= "0.18"}
+  "ounit2" {with-test}
+  "base-bytes"
+  "base-bigarray"
+  "base-unix"
+]
```
Comment (mseri) [inline packages/extunix/extunix.0.3.1/opam:44]:
```suggestion
  "ocaml" {>= "4.06"}
]
```
Looks like thr c stubs do not support older versions
```
#          gcc src/time.o (exit 1)
# (cd _build/default/src && /usr/bin/gcc -std=gnu99 -O2 -fno-strict-aliasing -fwrapv -fcommon -Wall -D_FILE_OFFSET_BITS=64 -D_REENTRANT -fPIC -std=gnu99 -O2 -fno-strict-aliasing -fwrapv -fcommon -Wall -D_FILE_OFFSET_BITS=64 -D_REENTRANT -fPIC -g -I /home/opam/.opam/4.04/lib/ocaml -I /home/opam/.opam/4.04/lib/bytes -o time.o -c time.c)
# time.c: In function 'caml_extunix_asctime':
# time.c:69:3: error: unknown type name 'char_os'; did you mean 'wchar_t'?
#    char_os buf[32]; /* user-supplied buffer which should have room for
#    ^~~~~~~
#    wchar_t
# time.c:80:30: warning: passing argument 2 of 'asctime_r' from incompatible pointer type [-Wincompatible-pointer-types]
#    if (NULL == asctime_r(&tm, buf)) {
#                               ^~~
# In file included from config.h:240,
#                  from time.c:5:
# /usr/include/time.h:149:14: note: expected 'char * restrict' but argument is of type 'int *'
#  extern char *asctime_r (const struct tm *__restrict __tp,
#               ^~~~~~~~~
# In file included from config.h:20,
#                  from time.c:5:
# time.c:84:14: warning: implicit declaration of function 'caml_copy_string_of_os'; did you mean 'caml_copy_string_array'? [-Wimplicit-function-declaration]
#    CAMLreturn(caml_copy_string_of_os(buf));
#               ^~~~~~~~~~~~~~~~~~~~~~
# /home/opam/.opam/4.04/lib/ocaml/caml/memory.h:340:29: note: in definition of macro 'CAMLreturnT'
#    type caml__temp_result = (result); \
#                              ^~~~~~
# time.c:84:3: note: in expansion of macro 'CAMLreturn'
#    CAMLreturn(caml_copy_string_of_os(buf));
#    ^~~~~~~~~~
# time.c: In function 'caml_extunix_strftime':
# time.c:91:3: error: unknown type name 'char_os'; did you mean 'wchar_t'?
#    char_os buf[256];
#    ^~~~~~~
#    wchar_t
```

**Comment by mseri** [comment]
Are the test failures on centos-7 and oraclelinux-7 expected?

```
#         test alias test/runtest (exit 1)
# (cd _build/default/test && ./test.exe)
# ....................F.............................
# ==============================================================================
# Failure: tests:9:pts
# 
# Unix_error : grantpt() : Invalid argument
# ------------------------------------------------------------------------------
# Ran: 40 tests in: 0.04 seconds.
# FAILED: Cases: 40 Tried: 40 Errors: 0 Failures: 1 Skip:  0 Todo: 0 Timeouts: 0.
# .
# ==============================================================================
# Failure: tests:9:pts
# 
# Unix_error : grantpt() : Invalid argument
# ------------------------------------------------------------------------------
# Ran: 40 tests in: 0.04 seconds.
# FAILED: Cases: 40 Tried: 40 Errors: 0 Failures: 1 Skip:  0 Todo: 0 Timeouts: 0.
# ....
# ==============================================================================
# Failure: tests:9:pts
# 
# Unix_error : grantpt() : Invalid argument
# ------------------------------------------------------------------------------
# Ran: 40 tests in: 0.04 seconds.
# FAILED: Cases: 40 Tried: 40 Errors: 0 Failures: 1 Skip:  0 Todo: 0 Timeouts: 0.
# 
# ==============================================================================
# Failure: tests:9:pts
# 
# Unix_error : grantpt() : Invalid argument
# ------------------------------------------------------------------------------
# Ran: 40 tests in: 0.04 seconds.
# FAILED: Cases: 40 Tried: 40 Errors: 0 Failures: 1 Skip:  0 Todo: 0 Timeouts: 0.
```
And
```
#         test alias test/runtest (exit 1)
# (cd _build/default/test && ./test.exe)
# ....................F................................
# ==============================================================================
# Failure: tests:9:pts
# 
# Unix_error : grantpt() : Invalid argument
# ------------------------------------------------------------------------------
# Ran: 40 tests in: 0.05 seconds.
# FAILED: Cases: 40 Tried: 40 Errors: 0 Failures: 1 Skip:  0 Todo: 0 Timeouts: 0.
# 
# ==============================================================================
# Failure: tests:9:pts
# 
# Unix_error : grantpt() : Invalid argument
# ------------------------------------------------------------------------------
# Ran: 40 tests in: 0.05 seconds.
# FAILED: Cases: 40 Tried: 40 Errors: 0 Failures: 1 Skip:  0 Todo: 0 Timeouts: 0.
```

File: packages/extunix/extunix.0.3.1/opam [hunk 2]
```diff
@@ -0,0 +1,68 @@
...
+build: [
+  ["dune" "subst"] {pinned}
```
Comment (mseri) [inline packages/extunix/extunix.0.3.1/opam:46]:
```suggestion
  ["dune" "subst"] {dev}
```

File: packages/extunix/extunix.0.3.1/opam (hunk 1 above)
Comment (ygrek) [reply packages/extunix/extunix.0.3.1/opam:45]: thanks, applied

File: packages/extunix/extunix.0.3.1/opam (hunk 2 above)
Comment (ygrek) [reply packages/extunix/extunix.0.3.1/opam]: applied

**Comment by ygrek** [comment]
will need to look into tests failure, but I will consider it not immediately blocking the release

**Comment by mseri** [comment]
Mmh, it appeared also on another test:
```
-         test alias test/runtest
- ...........................................
- Ran: 40 tests in: 0.04 seconds.
- OK
- .
- Internal error, please report upstream including the contents of _build/log.
- Description:
-   ("hooks failed",
-   { exns =
-       [ { exn =
-             "Unix.Unix_error(Unix.ENOENT, \"unlink\", \"/tmp/build2b91e7.dune/extunixac1694write\")"
-         ; backtrace =
-             "Raised by primitive operation at Stdune__path.clear_dir.(fun) in file \"src/stdune/path.ml\", line 1163, characters 15-34\n\
-              Called from Stdlib__list.iter in file \"list.ml\", line 110, characters 12-15\n\
-              Called from Stdune__path.clear_dir in file \"src/stdune/path.ml\" (inlined), line 1178, characters 20-45\n\
-              Called from Stdune__temp.clear_dir in file \"src/stdune/temp.ml\", line 84, characters 2-20\n\
-              Called from Stdune__exn_with_backtrace.try_with in file \"src/stdune/exn_with_backtrace.ml\", line 9, characters 8-12\n\
-              "
-         }
-       ]
-   })
- Raised at Stdune__code_error.raise in file "src/stdune/code_error.ml", line
-   9, characters 30-62
- Called from Stdlib.at_exit.(fun) in file "stdlib.ml", line 553, characters
-   62-65
- Called from Stdlib.do_at_exit in file "stdlib.ml" (inlined), line 556,
-   characters 20-39
- Called from Stdlib.exit in file "stdlib.ml", line 559, characters 2-15
- Called from Main in file "bin/main.ml", line 264, characters 11-17
- 
- I must not crash.  Uncertainty is the mind-killer. Exceptions are the
- little-death that brings total obliteration.  I will fully express my cases. 
- Execution will pass over me and through me.  And when it has gone past, I
- will unwind the stack along its path.  Where the cases are handled there will
- be nothing.  Only I will remain.
[ERROR] The compilation of extunix failed at "/home/opam/.opam/opam-init/hooks/sandbox.sh build dune build -p extunix -j 47 @install @runtest".
```

**Comment by mseri** [comment]
I have notified dune maintainers and confirmed that the issues is with dune being excessively strict when deleting temp files. I think this can be merged, @kit-ty-kate do you agree?

**Comment by avsm** [comment]
I filed a dune issue, looks good to merge!

**Comment by kit-ty-kate** [comment]
This release still had some issues. Our CI had some problems and did not test the revdeps without saying anything.
None of the versions of the `devkit` package compile. There seems to be some issues when linking:
```
# ocamlfind ocamlmktop -linkpkg -thread -thread -package yojson -package yojson -package libevent -package libevent -package netstring -package netstring -package lwt.unix -package lwt.unix -package extlib -package extlib -package pcre -package pcre -package zip -package zip -package curl -package curl -package curl.lwt -package curl.lwt -package extunix -package extunix -package bytes -package bytes control.cmo exn.cmo extEnum_merge.cmo extEnum.cmo factor.cmo lwt_engines.cmo prelude.cmo files.cmo logger.cmo stre.cmo devkit_ragel.cmo time.cmo log.cmo action.cmo nix.cmo async.cmo bit_struct_list.cmo cache.cmo extArg.cmo memory.cmo signal.cmo daemon.cmo httpev_common.cmo digest_auth.cmo exn_lwt.cmo mtq.cmo extThread.cmo fastBase64.cmo gzip_stream.cmo gzip_io.cmo htmlStream.cmo var.cmo web.cmo httpev.cmo idn.cmo pid.cmo logstash.cmo lwt_flag.cmo lwt_mark.cmo lwt_util.cmo mVar.cmo network.cmo parallel.cmo persist.cmo php_serialize.cmo reader.cmo static_config.cmo unsafeBitSet.cmo -o devkit.top
# + ocamlfind ocamlmktop -linkpkg -thread -thread -package yojson -package yojson -package libevent -package libevent -package netstring -package netstring -package lwt.unix -package lwt.unix -package extlib -package extlib -package pcre -package pcre -package zip -package zip -package curl -package curl -package curl.lwt -package curl.lwt -package extunix -package extunix -package bytes -package bytes control.cmo exn.cmo extEnum_merge.cmo extEnum.cmo factor.cmo lwt_engines.cmo prelude.cmo files.cmo logger.cmo stre.cmo devkit_ragel.cmo time.cmo log.cmo action.cmo nix.cmo async.cmo bit_struct_list.cmo cache.cmo extArg.cmo memory.cmo signal.cmo daemon.cmo httpev_common.cmo digest_auth.cmo exn_lwt.cmo mtq.cmo extThread.cmo fastBase64.cmo gzip_stream.cmo gzip_io.cmo htmlStream.cmo var.cmo web.cmo httpev.cmo idn.cmo pid.cmo logstash.cmo lwt_flag.cmo lwt_mark.cmo lwt_util.cmo mVar.cmo network.cmo parallel.cmo persist.cmo php_serialize.cmo reader.cmo static_config.cmo unsafeBitSet.cmo -o devkit.top
# File "_none_", line 1:
# Error: Required module `ExtUnix' is unavailable
# Command exited with code 2.
# make: *** [Makefile:39: top] Error 10
```

**Comment by mseri** [comment]
Argh. I had not yet realised the CI issue back then. There may actually be a number of other packages with broken revdeps

**Comment by kit-ty-kate** [comment]
Luckily this seems to be the only one for this PR: http://check.ocamllabs.io/diff/1617127610-86c6c9a7aa07f22c4aa560fa3003a920d2a8f13f..1618491516-7af4b295dc2de0b2b594c05ab4af40ef963bbc5c
There are two old packages also broken from the latest release of h2 but that's minor.

**Comment by ygrek** [comment]
it looks like only old versions of devkit are affected, it is not worth investigating why, I will just drop them with the next release of devkit.

**Comment by kit-ty-kate** [comment]
oops sorry my bad, indeed I missed devkit > 1.x for some reason when looking at their available versions. I'll send a PR to add the appropriate constraint to those old versions right away

**Pull request merged**